### DISCUSSION TITLE
I want to implement console support into a runtime

### DISCUSSION THREAD

**Discussion by nothendev** [tokio-rs/console#589] [Q&A] [answered]
How would I do it? There does not seem to be any docs about doing that in this repo or anywhere else.

**Comment by hds** [comment]
Do you mean you want to implement console support for some other runtime that isn't Tokio?

**Reply by nothendev** [reply]
exactly

**Reply by hds** [reply, marked as answer]
There is currently no specification for the instrumentation present in Tokio.

To understand it, you can read the `console-subscriber` source code or the `tokio` source code.

I have also written about it a bit on my personal blog in the following posts:
- https://hegdenu.net/posts/tracing-tokio-tasks/
- https://hegdenu.net/posts/debugging-tokio-instrumentation/
- https://hegdenu.net/posts/tokio-waker-instrumentation/
- https://hegdenu.net/posts/tracing-tokio-resources/

Certainly, starting with tasks, the `tokio` source code is quite clear in the fields that are included for task spawning:

https://github.com/tokio-rs/tokio/blob/512e9decfb683d22f4a145459142542caa0894c9/tokio/src/util/trace.rs#L54-L79

Down below that is the same for spawning blocking tasks.

To complete task instrumentation, you need to also instrument waker operations. You can see the format of that trace event in the `tokio` source here:

https://github.com/tokio-rs/tokio/blob/512e9decfb683d22f4a145459142542caa0894c9/tokio/src/runtime/task/waker.rs#L45-L55
